### ISSUE TITLE
Improve boot speed by eliminating unneeded icon file scans.

### ISSUE THREAD

**Issue by MarioG-X** [dakanji/RefindPlus#96]
### Suggested Feature

Instead of scanning every single partition and/or volume for volumebadge and volumeicon files at the start of RP, scan for those files only for the volumes that will be displayed.

### Background Information

Data only volumes and macOS "VM", "Preboot", "Update", and "Recovery" volumes are scanned. On the attached log **16 unneeded volumes are scanned for volumebadge, then 16 times again for volumeicon**, that's a lot of unneeded IO. Since none of these are displayed on the RP menu, why not check for those on the volume _only_ if the volume is added to the menu?

Imagine how long it takes when I have to boot from an external with 10 systems backed up on it.

I noticed EFI is also scanned but may be needed as a special case.

If this is a major technical change it may not be worthwhile if it may cause other issues. If it's doable you may want to consider it.

### Additional Context

Log:
[22d12v4316.log](https://github.com/dakanji/RefindPlus/files/8054449/22d12v4316.log)

**Comment by dakanji** [comment]
Thanks. 
Please use the various options provided in the config file to remove such unwanted icon scanning.

**Comment by MarioG-X** [comment]
Thanks for the suggestions, however I was only able to gain a little speed. Not a big deal unless I am testing RP a lot. Here are the results:

ignore_hidden_icons - Made the biggest difference. Unfortunately its all or nothing. The Elementary partition uses volumeicon and I was planning to use it to differentiate the levels of macOS. The log shows the more number of lines than volumebadge activity. This is the primary reason for the enhancement. Anyway, I don't want this off.

hideui badges - Was the second biggest difference, the log shows. Again, its all or nothing. The log shows the fewer number of lines than volumeicon activity. I don't use this so does not affect me.

scanfor - I boot from all options except manual so can't change this. I also removed default disabled stanzas, it made no perceptible difference.

showtools - I removed about half the functions, it made no perceptible difference.

I change this system a lot so don't want to have to update the config often using manual stanzas, not worth it.

No a big deal to live with it since it really only affects high volume testing. It's far worse when an external HD is connected but I don't do that often, only when testing your stuff (which reminds me, I need to test external with your last update!).

You can close this but if ignore_hidden_icons could be made to skip volumes not on the menu someday that would be nice. Thanks again!

**Comment by dakanji** [comment]
Revisited this and yes, the `VolumeIcon` icon searches could be optimised but the way it works, inherited from upstream, is as follows: 
- Every available volume instance is checked for VolumeIcons.
- This is done BEFORE it is determined whether the volume is to be shown or not
- If a VolumeIcon is found, this is stored in the an object containing info for the volume
- Later, when a volume is to be displayed, the VolumeIcon is used if previously found and there is no other icon for that volume.

That is the inherited inefficient search pattern from upstream.
What RefindPlus does extra, is to provide a `hidden_icons_prefer` token that will use the VolumeIcons once found but keeps the inherited search pattern that checks everything first. 
https://github.com/dakanji/RefindPlus/blob/5bc41be64f83aeb34ca8ba06bce46d848d409632/config.conf-sample#L283-L301

I don't use these icons and have not been motivated to look into changing the search pattern apart from that by default, there is a `HelpIcon` that disables such searches amongst other things unless `declined`
https://github.com/dakanji/RefindPlus/blob/5bc41be64f83aeb34ca8ba06bce46d848d409632/config.conf-sample#L46-L58

I always have this one but yes, not so good for those that want the VolumeIcons and other things that this feature disables.

Fixing this needs a bit of a big change to the processs flow to implement. Basically, disable the early seach and only search when needed. Didn't want to stray too far from the upsteam setup at the time but might take another look at this at some point as the implementation is a bit strange given that it only checks internal volumes. Added something to RefindPlus to also check external volumes if set.
https://github.com/dakanji/RefindPlus/blob/5bc41be64f83aeb34ca8ba06bce46d848d409632/config.conf-sample#L260-L266

Are you still noticing an impact from this?

**Comment by MarioG-X** [comment]
Sorry to take so long. I no longer have an issue with long RP menu time because I have eliminated the use of RP for macOS volumes, only used for Linux:

Mac mini - 
Now boot all macOSs on internal and external using the Option key. No longer use RP for booting macOSs.
Removed Linux from internal, boot all from external SSD using Option key to boot RP on external to boot any Linux.

iMac 2012 - 
Boots directly to Opencore to Catalina, Ventura, Sonoma, OC starts RP when Linux needed. All internal on HD and SSD.
